Support for Unknown Fields (Rebase of #574)

blizzardfinnegan opened the pull request:
For more information see Issue #2

This PR is largely similar to #574 , rebased to main, and with edits suggested by the maintainer on said PR implemented to the best of my ability.

Commit:
Blind copy of changes in PR #574

Does not compile, incomplete.

Commit:
Begin implementing necessary changes

- Rename UnknownFieldSet to UnknownFieldList, as requested by maintainer
  caspermeijn
- Change fixed32 and fixed64 representation to be consistent with C++
  implementation, as requested by caspermeijn
- start resolving build-time errors
    - Currently failing with reexported crate issues

Commit:
continue fixed32/fixed64 tweaks

- TODO: What endianness is protobuf?

Commit:
Clarify endianness of unknown fields

Commit:
Resolve minor testing issues

Commit:
Rename derive attribute

Commit:
Rename test messages to be more clear

Commit:
Temporarily comment out section

This section should be added back (and/or modified to fit new
standards), but commenting out for now to reduce points of failure to
resolve creating bad Copy derives

Commit:
Modify unknown fields to be Option

Instead of checking against empty, which can fail differently.
Attempting to force fields with unknown to not have a derived Copy.

Commit:
Update code-generator to support Option unknown

Commit:
Update name of unknown field in derive builds

Commit:
Fix bad derive Copy error

Commit:
Messages should be passed to config individually

Commit:
Finalise commits for first pass

Commit:
rustfmt and clippy

Commit:
Undo edit to otherwise untouched file

While this is a change made by rustfmt/clippy, it's outside the scope of
this PR, so it's been removed

Commit:
Resolve CI/clippy issue

Seems like local clippy check runs in rust edition 2021, CI runs in 2024

Commit:
Merge remote-tracking branch 'upstream/master' into feature/unknown_fields_support

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

Commit:
Confirm Protobuf unknown field tests pass

aidandj commented:
Ran into an issue using [pbjson_build](https://docs.rs/pbjson-build/latest/pbjson_build/) with this. Because when you generate unknown fields, pbjson_build doesn't know what to do with them and it errors:
```
error[E0063]: missing field `unknown_fields` in initializer of `google::protobuf::generated_code_info::Annotation`
    --> /out/pbjson/google.protobuf.serde.rs:3929:20
     |
3929 |                 Ok(super::tonic::google::protobuf::generated_code_info::Annotation {
```

caspermeijn commented on prost-derive/src/field/mod.rs:
```diff
@@ -86,6 +91,7 @@ impl Field {
             Field::Map(ref map) => vec![map.tag],
             Field::Oneof(ref oneof) => oneof.tags.clone(),
             Field::Group(ref group) => vec![group.tag],
+            Field::Unknown(_) => vec![],
```
Is this correct? The unknown fields have tags, right? I have not looked into what this function is used for.

caspermeijn commented on prost-derive/src/field/unknown.rs:
```diff
@@ -0,0 +1,66 @@
+
+        for attr in attrs {
+            if word_attr("unknown_fields", attr) {
+                set_bool(&mut unknown, "duplicate message attribute")?;
```
This error message is incorrect

caspermeijn commented on tests/src/decode_error.rs:
```diff
@@ -27,7 +27,10 @@ fn test_decode_error_multiple_levels() {
-            optional_foreign_message: Some(ForeignMessage { c: -1 }),
+            optional_foreign_message: Some(ForeignMessage {
+                c: -1,
+                ..Default::default()
```
This trick should be described in the user facing documentation

caspermeijn commented on conformance/failing_tests.txt:
```diff
@@ -1,3 +0,0 @@
-# TODO(tokio-rs/prost#2): prost doesn't preserve unknown fields.
```
Can this file be entirely deleted?

caspermeijn reviewed (changes requested):
Thank you for taking this forward. It is a large PR, so it takes some time for me to respond.

I miss user facing documentation. Could you add a section to the README that describes this new functionality?

Commit:
Implement some suggested changes

- Remove Option wrapper on unknown fields
- Actually utilise custom iterator instead of nested for loops
- use Iterator.sum() instead of manual counter
- use Buf.get_u64_le() and Buf.get_u32_le() instead of manual copies
- Update conformance to not require failing_tests.txt
- remove empty failing_tests.txt

Commit:
Continue with requested changes

- Add user-facing documentation for `prost` and `prost-build` for
  enabling unknown fields, along with limitations
- Modify `prost_build::Config::include_unknown_fields` to have a default
  value for the field in which unknown fields are stored. (defaults to
  `unknown_fields`)

blizzardfinnegan commented:
CI is currently failing, for reasons unclear to me. All CI failures are due to this error:
```
no function or associated item named `new` found for struct `error::DecodeError` in the current scope
```
Which is not accurate. It's only failing in CI (this failure does not occur if you run the same commands locally, afaict), and only on the files I've changed.

caspermeijn commented:
> Yes, that's what I used in my pbjson update, but either way, it's a breaking change if you want to support unknown Fields

No, this PR is not a breaking change. You have to explicitly enable the unknown fields, so existing users are not broken. Only the users that choose for unknown fields are broken.

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

Commit:
Resolve DecodeError CI error msg

Commit:
rustfmt && clippy

Commit:
Implement some changes suggested by @caspermeijn

Commit:
Remove custom iterator in favour of flat_map

blizzardfinnegan commented on prost/src/unknown.rs:
```diff
@@ -0,0 +1,172 @@
+
+    /// Gets an iterator over the fields contained in this set.
+    pub fn iter(&self) -> impl Iterator<Item = (u32, &UnknownField)> {
+        UnknownFieldIter {
```
Alright. After some more research, it is indeed possible, with the help of `core::iter::repeat()` and `Iterator::flat_map()`. If you think it needs more expansive documentation on what's going on under the hood, let me know and I'll update accordingly.

Commit:
Remove edge-case of multiple unknown fields

Commit:
Move unknown_fields to _unknown_fields to avoid collisions

blizzardfinnegan commented:
> Protobuf naming guide recommends not using underscores as the initial character.

This makes sense; Protobuf is language-agnostic, and some languages use underscores to denote private fields (python(?)), so useful to encourage people to not use it. Thank you for the information! I've updated accordingly.

caspermeijn commented on prost-build/src/lib.rs:
````diff
@@ -80,6 +80,40 @@
+//! ```rust,no_run
+//! use std::io::Result;
+//! fn main() -> Result<()> {
+//!     //prost_build::compile_protos(&["src/items.proto"], &["src/"])?;
````
This comment in the example is a bit weird

caspermeijn commented on prost-build/src/lib.rs:
```diff
@@ -80,6 +80,40 @@
+//! fn main() -> Result<()> {
+//!     //prost_build::compile_protos(&["src/items.proto"], &["src/"])?;
+//!     let mut config = prost_build::Config::new();
+//!     config.btree_map(["."]);
```
This line is not related to unknown fields

caspermeijn reviewed (changes requested):
I like the progress. I would like to see some small changes.

blizzardfinnegan commented on prost-derive/src/lib.rs:
```diff
@@ -108,21 +114,45 @@ fn try_message(input: TokenStream) -> Result<TokenStream, Error> {
+        let field_merge = field.merge(&prost_path, quote!(value));
+        merge.push(quote! {
+            _ => {
                 let mut value = &mut self.#field_ident;
```
It's not entirely clear? However, removing this line results in failing tests when using the `#[derive(::prost::Message)]` call in unit tests. 

```
cannot find value 'value' in this scope
    |
341 |     #[derive(Clone, PartialEq, Eq, Hash, ::prost::Message)]
    |                                          ^^^^^^^^^^^^^^^^ not found in this scope
    |
    = note: this error originates in the derive macro `::prost::Message`
```

There is a similar declaration in the merge field, and appears to be in line with the implementation in previous versions of the `merge` section.

Commit:
Update with requested changes

Commit:
Clarify doc comments regarding u32 and UnknownFields

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

caspermeijn commented:
> @caspermeijn Just checking in on the status of this MR; I know the past however long has been holidays, so projects and whatnot have probably taken a bit of a break, wanted to make sure it's still on the radar.

Yes, I see your progress. Thank you for that. It is difficult for me to find the time to review this large PR, as it needs a long block of time to fully understand the context. Please hang in there!

I did move one of your fixes of the bench to a seperate small PR ;-) https://github.com/tokio-rs/prost/pull/1404

Commit:
Remove change unrelated to PR intent

Moved to #1404 by maintainer

Commit:
Slightly clarify text of docs for prost and prost-build

Make clear the name of the unknown fields field in the generated code
_must_ be chosen. Resolve poor wording in prost-build docs.

DirectXMan12 commented on prost-build/src/config.rs:
```diff
@@ -296,7 +296,7 @@ impl Config {
     {
         let field_name: String = match override_field_name {
             Some(x) => x.as_ref().to_string(),
-            None => "unknown_fields".to_string(),
```
this is relatively minor, but since variables starting with `_` mean "ignored", i often see folks use fields starting with `_` to mean "never accessed" (e.g. a `PhantomData` field is often `_marker` or something similar), which makes this use a bit surprising.

while i understand the logic, this does make the rust code look pretty unidiomatic imo.  might be worth considering trailing underscore instead, which is pretty uncommon to use intentionally _except_ for avoid name collisions like this.

aidandj replied:
I believe I was the one who originally suggested the leading underscore. And it is because that is specifically called out as not recommended in the protobuf style guide. The style guide also calls out not using a trailing underscore. So this would fit the same purpose. (Just means I have more code to refactor 😆)

https://protobuf.dev/programming-guides/style/

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

Commit:
Tweak recommended name based on suggestions in issue

Underscore at end of name is more idiomatic for Rust; approved by
original suggestor of prepended underscore as acceptable alternative

Commit:
Resolve CI errors

- Fix failing test due to incomplete conversion
- revert change to prost-derive/src/field/unknown.rs that resulted in
  failing tests

blizzardfinnegan commented:
CI failures seem to be largely due to Github internal issues; will force a rerun tomorrow, hoping they (Github/Microslop) get their issues resolved.

Commit:
Force retrigger GitHub Actions

blizzardfinnegan committed:
Merge branch 'tokio-rs:master' into feature/unknown_fields_support

blizzardfinnegan commented:
@caspermeijn 
Been a while, just checking back in on the progress for this MR.
